components/feature/top-sites/src/main/java/mozilla/components/feature/top/sites/TopSitesMiddleware.kt

```diff
@@ -0,0 +1,128 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.top.sites
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.InitAction
+import mozilla.components.browser.state.action.TopSiteAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.TopSite
+import mozilla.components.browser.state.state.TopSite.Type.DEFAULT
+import mozilla.components.concept.storage.FrecencyThresholdOption
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.lib.state.Store
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * [Middleware] implementation for handling [TopSiteAction] and syncing the top sites in
+ * [BrowserState.topSites] with the [DefaultTopSitesStorage].
+ */
+class TopSitesMiddleware(
+    private val storage: Storage,
+    private val config: () -> TopSitesConfig,
+    coroutineContext: CoroutineContext = Dispatchers.IO
+) : Middleware<BrowserState, BrowserAction> {
+
+    private var scope = CoroutineScope(coroutineContext)
+
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        when (action) {
+            is InitAction -> initializeTopSites(context.store)
+            is TopSiteAction.AddTopSiteAction -> addTopSite(action)
+            is TopSiteAction.RemoveTopSiteAction -> removeTopSite(action)
+            else -> next(action)
+        }
+    }
+
+    private fun initializeTopSites(
+        store: Store<BrowserState, BrowserAction>
+    ) = scope.launch {
+        val (totalSites, frecencyConfig) = config.invoke()
+        val topSites = storage.getTopSites(
+            totalSites = totalSites,
+            frecencyConfig = frecencyConfig
+        )
+        store.dispatch(TopSiteAction.AddTopSitesAction(topSites))
+    }
+
+    private fun addTopSite(action: TopSiteAction.AddTopSiteAction) = scope.launch {
+        storage.addTopSite(
+            title = action.topSite.title,
+            url = action.topSite.url,
+            isDefault = action.topSite.type === DEFAULT
+        )
+    }
+
+    private fun removeTopSite(action: TopSiteAction.RemoveTopSiteAction) = scope.launch {
+        val (totalSites, frecencyConfig) = config.invoke()
+        val topSites = storage.getTopSites(
+            totalSites = totalSites,
+            frecencyConfig = frecencyConfig
+        )
+
+        // Find the matching top site in storage since top sites added from [AddPinnedSiteUseCase]
+        // does not contain the matching id between the Store and storage.
+        topSites.find {
```

Comment:
@pocmo This is a bit ugly because [`AddPinnedSiteUseCase`](https://github.com/mozilla-mobile/android-components/pull/11490/files#diff-7cb43b4c7ecfbfdc7e53e9ac90c94edd5efd61b80f361ee8df940a1dcced6e1aR30) will create a `TopSite`, which will not necessarily match what is added to storage because the `id` and `createdAt` are auto-generated at the storage level. Wondering if we have a better approach to resolving this so that we get a proper sync between the Store and Storage.

Reply:
I guess one solution would be to generate the top site `id` prior to adding to the storage. Only tricky bit is that we will need to probably change the db schema (probably drop `createdAt` while we are at it) and write a migration script, but not terrible.

Reply:
* `createdAt` is created at the storage level, but:
  * It doesn't have to. You could pass it in and locally add the timestamp whenever you create a `TopSite` without it (instead of `0` like you do now)
  * Or you could just set the creation timestamp. FWIW it may not matter if the one we store in the DB is a millisecond behind (if at all)? 🤔 
* For the ID:
  *  As you said, it doesn't need to be an auto_increment ID. You could instead use a UUID that you can generate yourself. That's how we do it when we are storing crashes in the crash reporter DB. Every crash gets a UUID.
  * Or you could backfill it. Once the middleware has saved the top site and knows the ID, it can dispatch an action to set the ID. Or you could even not forward the "add action" to the reducer (by not calling `next()`), save to the storage and then dispatch an action that adds the topsite with ID.